src/java/org/apache/cassandra/gms/Gossiper.java

```diff
@@ -2130,7 +2136,12 @@ public boolean waitForSchemaAgreement(long maxWait, TimeUnit unit, BooleanSuppli
 
     public boolean haveMajorVersion3Nodes()
     {
-        return haveMajorVersion3NodesMemoized.get();
+        return isUpgradingFromVersionLowerThan(4, 0);
+    }
+
+    public boolean isUpgradingFromVersionLowerThan(int major, int minor) {
+        CassandraVersion v = upgradeFromVersionMemoized.get();
+        return v != null && v.isLowerThan(major, minor);
```

Comment:
This is actually what I wanted - my reasoning was like - until we have no information about the cluster, we should assume the most legacy behaviour, which means that this method returns `true` for any reference version

Reply:
what happens when upgrading cluster from `4.0.0` to `4.0.1`? Do we expect `NULL_VERSION` being returned and doing the more expensive digest query during upgrade? The upgrade process in a large cluster can be time consuming.

Reply:
Given we use `isLowerThan(major, minor)`, the supplier will not detect any upgrade because `isLowerThan(4, 0)` returns false in all cases. So we only need the Gossiper to figure out versions of other nodes.

We start from `NULL_VERSION`, so we reduce the significantly digest mismatches during the phase when the Gossiper does not know the versions around the cluster. Another advantage of this approach is the monotonicity of the attribute we use to keep the current minimum version - once it is set to null (the cluster is consistent and stable) we avoid further checks. 

Disadvantage is that when we do not do the minor version upgrade (we do patch version upgrade or we just restart the cluster), we will run in legacy mode until the Gossiper detects versions of all the nodes.

Reply:
I do not worry about if the checker has gone to the step to check with `isLowerThan`. 

The concern comes from the line#188-189, where `NULL_VERSION` can be returned if the release version grabbed from peer is null. It can happen during any non-major upgrade. We should avoid such scenario.

```java
            if (version == null)
                return new ExpiringMemoizingSupplier.NotMemoized<>(SystemKeyspace.NULL_VERSION);
```

Reply:
So this is about which type of upgrades we want to optimise for during that ephemeral period - if for 3 -> 4, then we should stay with the current approach, if for 4 -> 4, we should just skip hosts with `null` version. 

I was wondering if there is a way to figure out whether the node has started for the first time with Cassandra 3 data (first started after upgrade). SSTable format has changed to `na` but I don't know if we force rewrite some system tables to the new format after startup. I'm just saying that if we could detect that the node has started for the first time after upgrading it from Cassandra 3, we could assume the current approach, otherwise we could skip null versions.

Reply:
so what do you think @yifan-c ?

Reply:
Regarding upgrade detection, I can think of 2 approaches. Both have their good and bad.
- Something similar to `SystemKeyspaceMigrator40`, that check if the new table introduced in 40 exist at local. The good: good UX. The bad: what if the instance has migrated the tables, but failed to start. The operator restarts the instance during cluster upgrade. Now the instance is unable to detect. 
- Accept a explicit property, say `cassandra.isUpgrade: Boolean`. The good: gives the operator more control. The bad: another property to handle (in some automation/manual process), and need to unset the property later.

Regarding returning `NULL_VERSION`, I just thought of a failure scenario...
The cluster is upgrading from `3.11` to `4.0`. When `NULL_VERSION` is returned, the 4.0 instance assumes the earliest behavior, which is `3.4-`, and there is a chance for digest mismatch. We can update the property to `cassandra.upgradeFrom: String = CassandraVersion`.

Reply:
I'm strongly against introducing additional user-facing property. Cassandra is difficult enough without that :)

I'll take it statistically - patch upgrades are much more frequent than major/minor version upgrades. Also I suppose the users are much more careful during major/minor version upgrades and probably less likely to exercise the cluster during that period. Therefore I'll optimise for that by just assuming the cluster is on 4.0, similarly as it has been done for 3.11